clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1800,11 +1888,186 @@ class SyclKernelDecompMarker : public SyclKernelFieldHandler {
+      : SyclKernelFieldHandler(S) {
+    // Generate new type
+    createNewType(RD);
+  }
+
+  bool enterStruct(const CXXRecordDecl *, FieldDecl *, QualType Ty) final {
+    // Generate new type
+    createNewType(Ty->getAsCXXRecordDecl());
+    return true;
+  }
+
+  bool leaveStruct(const CXXRecordDecl *, FieldDecl *FD, QualType Ty) final {
+    // Get Decl of generated new type
+    CXXRecordDecl *ModifiedRD = getGeneratedNewRecord(Ty->getAsCXXRecordDecl());
+
+    // Add this record as a field of it's parent record.
+    if (!ModifiedRecords.empty())
+      addField(FD, QualType(ModifiedRD->getTypeForDecl(), 0));
+    return true;
+  }
+
+  bool enterStruct(const CXXRecordDecl *, const CXXBaseSpecifier &,
+                   QualType Ty) final {
+    // Generate new type
+    createNewType(Ty->getAsCXXRecordDecl());
+    return true;
+  }
+
+  bool leaveStruct(const CXXRecordDecl *Parent, const CXXBaseSpecifier &BS,
+                   QualType Ty) final {
+    // Get Decl of generated new type
+    CXXRecordDecl *ModifiedRD = getGeneratedNewRecord(Ty->getAsCXXRecordDecl());
+
+    // Create CXXBaseSpecifier for this generated class.
+    createBaseSpecifier(Parent, ModifiedRD, BS);
+    return true;
+  }
+
+  bool handlePointerType(FieldDecl *FD, QualType FieldTy) final {
+    QualType PointeeTy = FieldTy->getPointeeType();
+    Qualifiers Quals = PointeeTy.getQualifiers();
+    auto AS = Quals.getAddressSpace();
+    // Leave global_device and global_host address spaces as is to help FPGA
+    // device in memory allocations
+    if (!PointeeTy->isFunctionType() && AS != LangAS::sycl_global_device &&
+        AS != LangAS::sycl_global_host)
+      Quals.setAddressSpace(LangAS::sycl_global);
+    PointeeTy = SemaRef.getASTContext().getQualifiedType(
+        PointeeTy.getUnqualifiedType(), Quals);
+    QualType ModTy = SemaRef.getASTContext().getPointerType(PointeeTy);
+    addField(FD, ModTy);
+    return true;
+    // We do not need to wrap pointers since this is a pointer inside
+    // non-decomposed struct.
+  }
+
+  bool handleScalarType(FieldDecl *FD, QualType FieldTy) final {
+    addField(FD, FieldTy);
+    return true;
+  }
+
+  bool handleUnionType(FieldDecl *FD, QualType FieldTy) final {
+    return handleScalarType(FD, FieldTy);
+  }
+
+  bool handleNonDecompStruct(const CXXRecordDecl *, FieldDecl *FD,
+                             QualType Ty) final {
+    addField(FD, Ty);
+    return true;
+  }
+
+  bool handleNonDecompStruct(const CXXRecordDecl *Parent,
+                             const CXXBaseSpecifier &BS, QualType Ty) final {
+    createBaseSpecifier(Parent, Ty->getAsCXXRecordDecl(), BS);
+    return true;
+  }
+
+  bool handleSimpleArrayType(FieldDecl *FD, QualType Ty) final {
+    addField(FD, Ty);
+    return true;
+  }
+
+  // FIXME: Array of pointers/ array of types containing pointers
+  // will be handled in a follow-up PR. Currently they continue to
+  // trigger decomposition.
```

Comment:
Hope we won't forget to remove all these comments when the support for arrays is implemented.

Reply:
I added comments wherever I removed array handling code which I had originally implemented in this PR, so that I could remember where all we needed to handle it. If you want, I can remove some of the FIXMEs if you prefer.

Reply:
> I added comments wherever I removed array handling code which I had originally implemented in this PR, so that I could remember where all we needed to handle it.

Ok, this makes sense.